source/delegate/contracts/interfaces/IDelegateV2.sol

```diff
@@ -0,0 +1,102 @@
+/*
+  Copyright 2020 Swap Holdings Ltd.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+
+pragma solidity 0.5.12;
+pragma experimental ABIEncoderV2;
+
+import "@airswap/types/contracts/Types.sol";
+
+interface IDelegateV2 {
+
+  struct Rule {
+    address senderToken;
+    address signerToken;
+    uint256 senderAmount;
+    uint256 signerAmount;
+    uint256 nextRuleID;
+    uint256 prevRuleID;
+  }
+
+  event CreateRule(
+    address indexed owner,
+    address indexed senderToken,
+    address indexed signerToken,
+    uint256 ruleID,
+    uint256 senderAmount,
+    uint256 signerAmount
+  );
+
+  event DeleteRule(
+    address indexed owner,
+    uint256 indexed ruleID
+  );
+
+  event ProvideOrder(
```

Comment:
So the important thing to remember when it comes to rule creation is that there's going to be a lot of need for lifecycle tracking. Lifecycle events include
- when a rule is created
- when a rule is deleted
- when a rule is filled against by an order

I think because of this, in cases where an order is filling against multiple rules, a ProvideOrder event will need to be emitted for each rule that an order fills against, with the senderAmount/signerAmount that each rule filled against the order. The ruleID will also need to be included in each provideOrder event. 

The provideOrder event is information for the delegate owner, not for the sender of the transaction. The delegate owner cares about the status of their individual rules, not the aggregate volume of a transaction against them.

Reply:
Yep that makes sense.

Do you think it would be better to have 2 events:
- RuleFilled(ruleID, senderAmount, signerAmount, etc)
- ProvideOrder(totalSenderAmount, totalSignerAmount, etc)

Is the combined ProvideOrder information in the above example needed? Or is just the RuleFilled information enough?

Reply:
I think all I need is the RuleFilled events.

Reply:
In addition, outside of just an event, is it worth storing information in a mapping to keep filled orders on the delegate?